pyomo/contrib/pynumero/interfaces/nlp_compositions.py

```diff
@@ -528,34 +456,38 @@ def evaluate_g(self, x, out=None, **kwargs):
         if out is None:
             res = self.create_vector_y()
         else:
-            raise NotImplementedError("ToDo")
+            assert isinstance(out, BlockVector), 'Composite NLP takes block vector to evaluate g'
+            assert out.nblocks == 2 * self.nblocks
+            assert out.size == self.ng
+            res = out
 
         if isinstance(x, BlockVector):
             assert x.size == self.nx
             assert x.nblocks == self.nblocks + 1
             for sid in range(self.nblocks):
+                # evaluate gi
                 self._nlps[sid].evaluate_g(x[sid], out=res[sid])
-                scenario_vids = self._zid_to_vid[sid]
-                diff = []
-                for zid in range(self.nz):
-                    vid = scenario_vids[zid]
-                    diff.append(x[sid][vid] - x[self.nblocks][zid])
-                res[sid + self.nblocks] = np.array(diff, dtype=np.double)
+
+                # evaluate coupling Ax-z
+                A = self._AB_csr[sid, sid]
+                res[sid + self.nblocks] = A * x[sid] - x[self.nblocks]
+                # scenario_vids = self._zid_to_vid[sid]
+                # diff = []
+                # for zid in range(self.nz):
+                #     vid = scenario_vids[zid]
+                #     diff.append(x[sid][vid] - x[self.nblocks][zid])
+                # res[sid + self.nblocks] = np.array(diff, dtype=np.double)
```

Comment:
Should this commented code be removed?  Otherwise, please leave a comment as to why it's here.

Reply:
Removed